File: src/Soros.js
function Soros(program, lang) {
    this.funcpat = /(\|?(\uE008\()+)?(\|?\uE008\(([^\(\)]*)\)\|?)(\)+\|?)?/
    this.meta = "\\\"$()|#;[]"
    this.enc = "\uE000\uE001\uE002\uE003\uE004\uE005\uE006\uE007\uE008\uE009"
    this.lines = []
    if (!/__numbertext__/.test(program))
        program = "__numbertext__;" + program

    program = program.replace("__numbertext__",
        // default left zero deletion
        "\"([a-z][-a-z]* )?0+(0|[1-9]\\d*)\" $(\\1\\2);" +
        // separator function
        "\"\uE00A(.*)\uE00A(.+)\uE00A(.*)\" \\1\\2\\3;" +
        // no separation, if subcall returns with empty string
        "\"\uE00A.*\uE00A\uE00A.*\"")

    // subclass for line data
    this.linetype = function (regex, repl, begin, end) {
        this.pat = regex
        this.repl = repl
        this.begin = begin
        this.end = end
    };

    // strip function
    this.strip = function (st, ch) {
        if (st == undefined) return ""
        return st.replace(new RegExp("^" + ch + "+"), "")
            .replace(new RegExp(ch + "+$"), "")
    };

    // character translation function
    this.tr = function (text, chars, chars2, delim) {
        for (var i = 0; i < chars.length; i++) {
            var s = delim + chars[i]
            while (text.indexOf(s) >= 0) {
                text = text.replace(s, chars2[i]);
            }
        }
        return text
    };

    // private run function
    this._run = function (data, begin, end) {
        for (var i in this.lines) {
            var l = this.lines[i]
            if (! ((!begin && l.begin) || (!end && l.end))) {
                var m = l.pat.exec(data)
                if (m != null) {
                    var s = data.replace(l.pat, l.repl)
                    var n = this.funcpat.exec(s)
                    while (n != null) {
                        var b = false
                        var e = false
                        if (n[3][0] == "|" || n[0][0] == "|") {
                            b = true
                        } else if (n.index == 0) {
                            b = begin
                        }
                        if (n[3][n[0].length - 1] == "|" || n[3][n[0].length - 1] == "|") {
                            e = true
                        } else if (n.index + n[0].length == s.length) {
                            e = end
                        }
                        s = s.substring(0, n.index + (n[1] == undefined ? 0 : n[1].length)) + this._run(n[4], b, e) +
                            s.substring(n.index + (n[1] == undefined ? 0 : n[1].length) + n[3].length)
                        n = this.funcpat.exec(s)
                    }
                    return s
                }
            }
        }
        return ""
    };

    // run with the string input parameter
    this.run = function (data) {
        data = this._run(this.tr(data, this.meta, this.enc, ""), true, true)
        return this.tr(data, this.enc, this.meta, "")
    };

    // constructor
//    program = program.replace(/\\\\/g, "\uE000")
//    program = program.replace(/\\[(]/g, "\uE003")
//    program = program.replace(/\\[)]/g, "\uE004")
//    program = program.replace(/\\[|]/g, "\uE005")
    program = this.tr(program, this.meta, this.enc, "\\")
    // switch off all country-dependent lines, and switch on the requested ones
    program = program.replace(/(^|[\n;])([^\n;#]*#[^\n]*[[]:[^\n:\]]*:][^\n]*)/g, "$1#$2")
        .replace(new RegExp("(^|[\n;])#([^\n;#]*#[^\n]*[[]:" + lang.replace("_", "-") + ":][^\n]*)", "g"), "$1$2")
    var l = program.replace(/(#[^\n]*)?(\n|$)/g, ";").split(";")
    var matchline = new RegExp(/^\s*(\"[^\"]*\"|[^\s]*)\s*(.*[^\s])?\s*$/)
    var prefix = ""
    for (var i in l) {
        var macro = /== *(.*[^ ]?) ==/.exec(l[i])
        if (macro != null) {
            prefix = macro[1]
            continue
        }
        var s = matchline.exec(l[i])
        if (prefix != "" && l[i] != "" && s != null) {
            s1 = this.strip(s[1], "\"")
            var empty = (s1 == "")
            var start = (!empty && s1[0] == '^')
            if (s[2] == undefined) s[2] = ""
            l2 = "\"" + (start ? "^" : "") + prefix + (empty ? "" : " ") +
                          s1.replace("^\^", "") + "\" " + s[2]
            s = matchline.exec(l2)
        }
        if (s != null) {
            s[1] = this.strip(s[1], "\"")
            if (s[2] == undefined) s[2] = ""; else s[2] = this.strip(s[2], "\"")
            var line = new this.linetype(
                new RegExp("^" + s[1].replace("^\^", "").replace("\$$", "") + "$"),
                s[2].replace(/\\n/g, "\n")
                    // call inner separator: [ ... $1 ... ] -> $(\uE00A ... \uE00A$1\uE00A ... )
                    .replace(/^[[]\$(\d\d?|\([^\)]+\))/g,"$(\uE00A\uE00A|$$$1\uE00A")
                    .replace(/[[]([^\$[\\]*)\$(\d\d?|\([^\)]+\))/g,"$(\uE00A$1\uE00A$$$2\uE00A")
                    .replace(/\uE00A]$/, "|\uE00A)") // add "|" in terminating position
                    .replace(/]/g, ")")
                    .replace(/(\$\d|\))\|\$/g,"$1||$$") // $(..)|$(..) -> $(..)||$(..)
                    .replace(/\$/g, "\uE008")
                    .replace(/\\0/g, "$$&")
                    .replace(/\\(\d)/g, "$$$1")
                    .replace(/\uE008(\d)/g, "\uE008($$$1)"),
                /^\^/.test(s[1]), /\$$/.test(s[1])
            )
            this.lines = this.lines.concat(line)
        }
    }
};

File: src/Soros.py
"Soros interpreter (see http://numbertext.org)"
from __future__ import unicode_literals
from __future__ import print_function
import re, sys

def run(program, data):
    return compile(program).run(data)

def compile(program):
    return _Soros(program)

# conversion function
def _tr(text, chars, chars2, delim):
    for i in range(0, len(chars)):
        text = text.replace(delim + chars[i], chars2[i])
    return text

# string literals for metacharacter encoding
_m = "\\\";#$()|[]"
_c = u"\uE000\uE001\uE002\uE003\uE004\uE005\uE006\uE007\uE008\uE009" # Unicode private area
_pipe = u"\uE003"
# separator prefix = \uE00A

# pattern to recognize function calls in the replacement string
_func = re.compile(_tr(r"""(?:\|?(?:\$\()+)?  # optional nested calls
                (\|?\$\(([^\(\)]*)\)\|?)      # inner call (2 subgroups)
                (?:\)+\|?)?""",               # optional nested calls
                _m[4:8], _c[:4], "\\"), re.X)  # \$, \(, \), \| -> \uE000..\uE003

class _Soros:
    def __init__(self, prg):
        self.lines = []
        if prg.find("__numbertext__") == -1:
            prg = "__numbertext__;" + prg
        # default left zero deletion
        # and separator function (no separation, if subcall returns with empty string)
        prg = prg.replace("__numbertext__", u"""0+(0|[1-9]\\d*) $1
\"([a-z][-a-z]* )0+(0|[1-9]\\d*)\" $(\\1\\2)
\"\uE00A(.*)\uE00A(.+)\uE00A(.*)\" \\1\\2\\3
\"\uE00A.*\uE00A\uE00A.*\"
""")
        prg = _tr(prg, _m[:4], _c[:4], "\\") # \\, \", \;, \# -> \uE000..\uE003
        matchline = re.compile("^\s*(\"[^\"]*\"|[^\s]*)\s*(.*[^\s])?\s*$")
        prefix = ""
        for s in re.sub("(#[^\n]*)?(\n|$)", ";", prg).split(";"):
            macro = re.match("== *(.*[^ ]?) ==", s)
            if macro != None:
                prefix = macro.group(1)
                continue
            m = matchline.match(s)
            if prefix != "" and s != "" and m != None:
                s = m.group(1).strip("\"")
                space = " " if s != "" else ""
                caret = ""
                if s[0:1] == "^":
                    s = s[1:]
                    caret = "^"
                s2 = m.group(2) if m.group(2) != None else ""
                s = "\"" + caret + prefix + space + s + "\" " + s2
                m = matchline.match(s)
            if m != None:
                s = _tr(m.group(1).strip("\""), _c[1:4], _m[1:4], "") \
                    .replace(_c[_m.find("\\")], "\\\\") # -> \\, ", ;, #
                if m.group(2) != None:
                    s2 = m.group(2).strip("\"")
                else:
                    s2 = ""
                s2 = _tr(s2, _m[4:], _c[4:], "\\") # \$, \(, \), \|, \[, \] -> \uE004..\uE009
                # call inner separator: [ ... $1 ... ] -> $(\uE00A ... \uE00A$1\uE00A ... )
                s2 = re.sub(r"[[]\$(\d\d?|\([^\)]+\))",u"$(\uE00A\uE00A|$\\1\uE00A", s2)
                s2 = re.sub(r"[[]([^\$[\\]*)\$(\d\d?|\([^\)]+\))",u"$(\uE00A\\1\uE00A$\\2\uE00A", s2)
                s2 = re.sub(r"\uE00A]$","|\uE00A)", s2) # add "|" in terminating position
                s2 = re.sub(r"]",")", s2)
                s2 = re.sub(r"(\$\d|\))\|\$", r"\1||$", s2) # $()|$() -> $()||$()
                s2 = _tr(s2, _c[:4], _m[:4], "")   # \uE000..\uE003-> \, ", ;, #
                s2 = _tr(s2, _m[4:8], _c[:4], "")   # $, (, ), | -> \uE000..\uE003
                s2 = _tr(s2, _c[4:], _m[4:], "") # \uE004..\uE009 -> $, (, ), |, [, ]
                s2 = re.sub(r"\\(\d?\d)", r"\\g<\1>",
                    re.sub(r"\uE000(\d?\d)", "\uE000\uE001\\\\g<\\1>\uE002", s2))
                try:
                    self.lines = self.lines + [[
                        re.compile("^" + s.lstrip("^").rstrip("$") + "$"),
                        s2, s[:1] == "^", s[-1:] == "$"]]
                except:
                    print("Error in following regex line: " + s, file=sys.stderr)
                    raise

    def run(self, data):
        return self._run(data, True, True)

    def _run(self, data, begin, end):
        for i in self.lines:
            if not ((begin == False and i[2]) or (end == False and i[3])):
                m = i[0].match(data)
                if m:
                    try:
                        s = m.expand(i[1])
                    except:
                        print("Error for the following input: " + data, file=sys.stderr)
                        raise
                    n = _func.search(s)
                    while n:
                        b = False
                        e = False
                        if n.group(1)[0:1] == _pipe or n.group()[0:1] == _pipe:
                            b = True
                        elif n.start() == 0:
                            b = begin
                        if n.group(1)[-1:] == _pipe or n.group()[-1:] == _pipe:
                            e = True
                        elif n.end() == len(s):
                            e = end
                        s = s[:n.start(1)] + self._run(n.group(2), b, e) + s[n.end(1):]
                        n = _func.search(s)
                    return s
        return ""
